astropy/io/fits/hdu/hdulist.py

```diff
@@ -512,6 +510,15 @@ def fileinfo(self, index):
 
         return output
 
+    def _load_all_hdus(self):
+        while self._read_next_hdu():
+            pass
+
+    def pop(self, index=-1):
+        # Make sure that HDUs are loaded before attempting to pop
+        self._load_all_hdus()
+        super(HDUList, self).pop(index)
```

Comment:
Another option would be to use `__getitem__`, to avoid loading all hdus if it is not necessary. `MutableSet` (which should be inherited instead of `list`, but there is probably a good reason why it is not used :thinking: ) does something like this:
```
    def pop(self, index=-1):
        '''S.pop([index]) -> item -- remove and return item at index (default last).
           Raise IndexError if list is empty or index is out of range.
        '''
        v = self[index]
        del self[index]
        return v
```
Not sure if there is a drawback if using this.

Reply:
That's an interesting idea -- wouldn't it require changes to a lot of other methods that are currently inherited from `list`, though?

Reply:
You mean if inheriting from `MutableSequence` ? It should not change that much I think, but you never know.

Reply:
Oh okay, that makes more sense. At first I thought it said `MutableSet`. Sounds like it's worth investigating, but maybe not in this PR?

Reply:
Yeah sorry I wrote the wring class :wink: 
It better to keep this for another indeed, as this one should be easy to backport if we want to.